File: languages/div.rb
class DivError < StandardError
  def initialize(message = "division by zero")
    super(message)
  end
end

def div(x, y)
  raise DivError if y == 0.0
  x / y
end

def main
  x, y = 1.0, 0.0
  
  begin
    z = div(x, y)
    puts "The result is: #{z}"
  rescue DivError => e
    warn "Division failed: #{e.message}"
  end
end

main if __FILE__ == $0

File: languages/div.pl
#!/usr/bin/perl
use strict;
use warnings;

package DivError;
use base 'Exporter';
our @EXPORT = qw(DivError);

sub new {
    my ($class, $message) = @_;
    $message //= "division by zero";
    return bless { message => $message }, $class;
}

sub message {
    my $self = shift;
    return $self->{message};
}

package main;

sub div {
    my ($x, $y) = @_;
    if ($y == 0) {
        die DivError->new();
    }
    return $x / $y;
}

sub main {
    my ($x, $y) = (1.0, 0.0);
    
    eval {
        my $z = div($x, $y);
        print "The result is: $z\n";
    };
    
    if ($@) {
        if (ref($@) eq 'DivError') {
            print STDERR "Division failed: " . $@->message() . "\n";
        } else {
            print STDERR "Division failed: division by zero\n";
        }
    }
}

main();